pytorch_lightning/trainer/trainer_conf.py

```diff
@@ -0,0 +1,259 @@
+)
+
+
+@dataclass
+class MLFlowLoggerConf:
+    experiment_name: str = "default"
+    tracking_uri: Optional[str] = None
+    tags: Optional[Dict[str, Any]] = None
+    save_dir: Optional[str] = None
+
+
+cs.store(
+    group="pl/logger",
+    name="mlflow",
+    node=ObjectConf(cls="pytorch_lightning.loggers.mlflow.MLFlowLogger", params=MLFlowLoggerConf()),
+)
+
+
+@dataclass
+class NeptuneLoggerConf:
+    api_key: Optional[str] = None
+    project_name: Optional[str] = None
+    close_after_fit: Optional[bool] = True
+    offline_mode: bool = False
+    experiment_name: Optional[str] = None
+    upload_source_files: Optional[List[str]] = None
+    params: Optional[Dict[str, Any]] = None
+    properties: Optional[Dict[str, Any]] = None
+    tags: Optional[List[str]] = None
+
+
+cs.store(
+    group="pl/logger",
+    name="neptune",
+    node=ObjectConf(cls="pytorch_lightning.loggers.neptune.NeptuneLogger", params=NeptuneLoggerConf()),
+)
+
+
+@dataclass
+class TensorboardLoggerConf:
+    save_dir: str = ""
+    name: Optional[str] = "default"
+    version: Any = None
+
+
+cs.store(
+    group="pl/logger",
+    name="tensorboard",
+    node=ObjectConf(cls="pytorch_lightning.loggers.tensorboard.TensorBoardLogger", params=TensorboardLoggerConf()),
+)
+
+
+@dataclass
+class TestTubeLoggerConf:
+    save_dir: str = ""
+    name: str = "default"
+    description: Optional[str] = None
+    debug: bool = False
+    version: Optional[int] = None
+    create_git_tag: bool = False
+
+
+cs.store(
+    group="pl/logger",
+    name="testtube",
+    node=ObjectConf(cls="pytorch_lightning.loggers.test_tube.TestTubeLogger", params=TestTubeLoggerConf()),
+)
+
+
+@dataclass
+class WandbConf:
+    name: Optional[str] = None
+    save_dir: Optional[str] = None
+    offline: bool = False
+    id: Optional[str] = None
+    anonymous: bool = False
+    version: Optional[str] = None
+    project: Optional[str] = None
+    tags: Optional[List[str]] = None
+    log_model: bool = False
+    experiment = None
+    entity = None
+    group: Optional[str] = None
+
+
+cs.store(
+    group="pl/logger",
+    name="wandb",
+    node=ObjectConf(cls="pytorch_lightning.loggers.wandb.WandbLogger", params=WandbConf()),
+)
+
+
+@dataclass
+class TrainerIn(LightningTrainerConf):
+    logger: Optional[Any] = None
+    profiler: Optional[Any] = None
+    checkpoint_callback: Optional[Any] = None
+    early_stop_callback: Optional[Any] = None
+    callbacks: Optional[Any] = None
+
+
+def init_trainer(cfg):
+    def conf_init(cfg, key):
+        return instantiate(getattr(cfg, key)) if hasattr(cfg, key) else None
+
+    profiler = conf_init(cfg.pl, "profiler")
+    early_stopping = conf_init(cfg.pl, "early_stopping")
+    checkpoint = conf_init(cfg.pl, "checkpoint")
+    logger = conf_init(cfg.pl, "logger")
+    if cfg.callbacks_list:
+        callbacks = [instantiate(callback) for callback in cfg.callbacks_list]
+    else:
+        callbacks = []
+    return asdict(
```

Comment:
why do you need to use datacalass.asdict here?

Reply:
I think I don't like this function:
it hides things that I users should be aware of to make good use of Hydra.

One thing you are missing is that you are not really initializing your optimizer:
it needs a passthrough of the model parameters, which is a runtime thing.
I don't think you are doing it in this example.

Reply:
I configure the optimizer in the lightning module in the configure optimizer function, where i give the params passthrough. I agree this function does hide functionality, do you have any suggestions to rework? At the moment, I think this is something the user could better design based on their scenario.

Reply:
asdict may not be necessary but we do have to passthrough the Hydra trainer arguments as separate keyword arguments, i dont think trainer supports OmegaConf.

Reply:
PL has started to support OmegaConf, I am not sure if the trainer supports it but let's try.
worse case we should fix it (I expect it to be pretty light and uncontroversial changes).

> I agree this function does hide functionality, do you have any suggestions to rework? At the moment, I think this is something the user could better design based on their scenario.

As a first step, let's move the logic back to user's main.
it's going to be a bit messier but it will allow us to see what should change to make this clean more clearly.

Reply:
About configure optimizer:
I am not seeing it in this diff. did it get lost somehow?

Reply:
Oh, configure_optimizers(), got it.
who is supposed to be calling it?

Reply:
Well, the usual context is that, you configure the optimizer inside that function in the Pytorch Lightning module. So, my setup for now has been creating all these configs for each optimizer, then let the user passthrough the parameters for the optimizer.

Reply:
Back to the asdict, right now for the trainer arguments configuration, I have a class LightningTrainerConf which is used as a structured config, I then extend this data class to TrainerIn which is the arguments I am passing to the Trainer object and finally call asdict and expand into keyword arguments. This flow is somewhat cumbersome, this is the class we are configuring: https://github.com/PyTorchLightning/pytorch-lightning/blob/96b32bee04c0dc971323bccc02f748b63f0ba09a/pytorch_lightning/trainer/trainer.py#L135. We would need to create a PR into PL to better support OmegaConf for trainer.

Reply:
okay, i cleaned this up in a recent commit.

Reply:
> I have a class LightningTrainerConf which is used as a structured config

Let's be precise:
You have LightningTrainerConf, you are constructing an DictConfig from it that looks like it. you do not operate on an instance of LightningTrainerConf  when you call as_dict, but on an instance if DictConfig.
I am somewhat surprised worked.